### PR TITLE
Update inference tests to 0.4.0 and add colocated driver capability

### PR DESCRIPTION
**Pull request by Spartee**
With the addition of the db colocation feature in SmartSim, the scaling tests needed to be updated to be able to run colocated models.

The client application can now set the model amongst other inference settings at runtime if specified in the environment of client application through the driver script. 

TODO
 - [x] Update client application to SmartRedis 0.3.0
 - [x] cpp-throughput tests should be updated 
 - [x] add instructions for using KeyDB instead of Redis (ticket made)
 - [x] provide tuning instructions and add to benchmarking results. (ticket made)

### CODE REVIEW

File: cpp-inference/CMakeLists.txt [hunk 1]
```diff
@@ -3,7 +3,7 @@ project(CppInferenceTests)
 cmake_minimum_required(VERSION 3.10)
 
 # set this to the path to your smartredis installation
-set(SMARTREDIS ../smartredis-0.1.1/)
+set(SMARTREDIS /home/users/spartee/dev/SmartRedis)
```
Comment (ben-albrecht) [inline cpp-inference/CMakeLists.txt:6]: unintentional local path?

File: cpp-inference/inference_scaling_imagenet.cpp [hunk 2]
```diff
@@ -69,34 +128,55 @@ void run_mnist(const std::string& model_name,
...
+  //timing_file << rank << "," << "unpack_tensor" << ","
+  //            << delta_t << std::endl << std::flush;
   }
   double loop_end = MPI_Wtime();
   delta_t = loop_end - loop_start;
+
+  // write times to file
+  for (int i=0; i<100; i++) {
+
+    timing_file << rank << "," << "put_tensor" << ","
```
Comment (mellis13) [inline cpp-inference/inference_scaling_imagenet.cpp:167]: To put less stress on file writes for large scaling studies, I wonder if we should take out the ``std::flush`` calls in this loop.  (I actually think that ``std::endl`` flushes the data stream so we could take the ``std::endl`` out of this loop completely and replace with ``\n`` if we wanted be more efficient with file writes.

File: cpp-inference/inference_scaling_imagenet.cpp [hunk 3]
```diff
@@ -69,34 +128,55 @@ void run_mnist(const std::string& model_name,
                       SmartRedis::MemoryLayout::nested);
     double put_tensor_end = MPI_Wtime();
     delta_t = put_tensor_end - put_tensor_start;
-    timing_file << rank << "," << "put_tensor" << ","
-                << delta_t << std::endl << std::flush;
+    put_tensor_times.push_back(delta_t);
+    //timing_file << rank << "," << "put_tensor" << ","
```
Comment (mellis13) [inline cpp-inference/inference_scaling_imagenet.cpp:132]: I like that we have moved the timings outside of the loop.  Maybe we can remove these old comments?

File: cpp-inference/inference_scaling_imagenet.cpp [hunk 4]
```diff
@@ -69,34 +128,55 @@ void run_mnist(const std::string& model_name,
...
-              << delta_t << std::endl << std::flush;
+  unpack_tensor_times.push_back(delta_t);
+  //timing_file << rank << "," << "unpack_tensor" << ","
+  //            << delta_t << std::endl << std::flush;
   }
   double loop_end = MPI_Wtime();
   delta_t = loop_end - loop_start;
+
+  // write times to file
+  for (int i=0; i<100; i++) {
```
Comment (mellis13) [inline cpp-inference/inference_scaling_imagenet.cpp:165]: Maybe we define a constant for the number of trials now that we use the constant (e.g. ``100``) in two loops?

**Review by mellis13 — APPROVED** [review]
Really nice upgrades to the scaling test.  I added a couple of comments for potential optimizations, but otherwise looks great.

**Review by billschereriii — CHANGES_REQUESTED** [review]
Enumeration values are not updated, so this branch does not compile with SmartRedis 0.4.0

**Comment by Spartee** [comment]
Throughput and inference tests are both now updated and work with the same result processing functions in the driver.

the processing of results was split off into its own CLI function and provides nice features like a loading bar 
Screen Shot 2022-03-15 at 7 14 10 PM

the result processing can be pointed at any dir with runs in it and it will create a results folder. you'll also notice that a log of the entire scaling run will be saved by the day it was executed. many successive runs can occur without overwriting results.
Screen Shot 2022-03-17 at 3 21 52 PM

The results folder will now how a total csv for all runs. and each run will have it's own csv record. if you have only processed some of the runs, you can avoid processing them again with the ``--overwrite=False`` flag as seen in the README. Each of the results dirs will be by the name of the run and have the created plots saved within it.
Screen Shot 2022-03-17 at 3 23 32 PM

Lastly, each run with data will have a ``run.cfg`` file for later reference that contains information about the run. this data will also be present in the CSV files.

The config looks like

```ini
[run]
name = infer-sess-colo-N12-T16-DBN12-DBC4-DBTPQ2-8332
path = /lus/cls01029/spartee/poseidon/smartsim-scaling/inference-scaling/infer-sess-colo-N12-T16-DBN12-DBC4-DBTPQ2-8332
smartsim_version = 0.4.0
smartredis_version = 0.3.0
db = keydb-server
date = 2022-03-17

[attributes]
colocated = 1
pin_app_cpus = 1
client_total = 192
client_per_node = 16
client_nodes = 12
database_nodes = 12
database_cpus = 4
database_threads_per_queue = 2
batch_size = 1
device = GPU
num_devices = 1
```

File: cpp-inference/CMakeLists.txt (hunk 1 above)
Comment (billschereriii) [reply cpp-inference/CMakeLists.txt:6]: yeah, let's make this path ../smartredis as the default

File: cpp-inference/inference_scaling_imagenet.cpp [hunk 5]
```diff
@@ -2,6 +2,47 @@
...
+bool get_cluster_flag() {
+  char* cluster_flag = std::getenv("SS_CLUSTER");
+  bool use_cluster = cluster_flag ? std::stoi(cluster_flag) : false;
+  return use_cluster;
+}
+
+int get_client_count() {
+  char* client_count_flag = std::getenv("SS_CLIENT_COUNT");
+  int client_count = client_count_flag ? std::stoi(client_count_flag) : 18;
+  return client_count;
```
Comment (billschereriii) [inline cpp-inference/inference_scaling_imagenet.cpp:42]: Recommend making a helper routine:

```
int get_env_integer(const char *var_name, int default)
{
   char *var_str = std::getenv(var_name);
   return (var_str == NULL) ? default : std::stoi(var_str);
}
```
With that, get_client_count (for example) becomes:

```
int get_client_count() {
  return get_env_integer("SS_CLIENT_COUNT", 18);
}
```

Comment (billschereriii) [reply cpp-inference/inference_scaling_imagenet.cpp:42]: Also, 18 is a weird default -- please add a comment explaining where that comes from

**Comment by billschereriii** [comment]
Two more comments:
1. Rather than explicitly saying to install fire, maybe we should make a `requirements.txt` and tell users to pip install from it?
2. Can we make default run.sh scripts for horizon, crystal, osprey?

**Comment by Spartee** [comment]
@billschereriii 

> Rather than explicitly saying to install fire, maybe we should make a requirements.txt and tell users to pip install from it?

The readme "explicitly" states to install from requirements.txt

> Can we make default run.sh scripts for horizon, crystal, osprey?

``batch_scripts`` folder has batch script examples for each of the tests for horizon. easily adapted to osprey. rest are out of scope here. most likely to be added to #9 

>  Also, 18 is a weird default -- please add a comment explaining where that comes from

this is from core count on horizon. As all of the tests set this from the driver, the default is never used. It's there solely for consistency.

helper routine is a good idea. thatll clean things up.

**Comment by billschereriii** [comment]
> > Rather than explicitly saying to install fire, maybe we should make a requirements.txt and tell users to pip install from it?
> 
> The readme explicitly states to install from requirements.txt

My bad, I think I was looking at an out of date readme.md

**Pull request merged**